keras_nlp/layers/bert_packer.py

```diff
@@ -0,0 +1,230 @@
+    output sequence will always start with a `start_value` and contain an
+    `end_value` after each sequence.
+
+    If inputs are batched, inputs should be `tf.RaggedTensor`s with shape
+    `[batch_size, None]` and will be packed and converted to a dense tensor with
+    shape `[batch_size, sequence_length]`.
+
+    If inputs are unbatched, inputs should be dense rank-1 tensors of any shape,
+    and will be packed to shape `[sequence_length]`.
+
+    Returns a python dictionary with three elements:
+      - `"tokens"`: The packed token `tf.Tensor`.
+      - `"padding_mask"`: A `tf.Tensor` with the same shape as `"tokens"`,
+        containing 0s where inputs have been padded and 1s elsewhere.
+      - `"segment_ids"`: A `tf.Tensor` with the same shape as `"tokens"`,
+        containing an int id maching tokens to the index of input sequence they
+        belonged to.
+
+    Args:
+        sequence_length: The desired output length.
+        start_value: The id or token that is to be placed at the start of each
+            sequence (called "[CLS]" for BERT). The dtype much mach the dtype of
+            the input tensors to the layer.
+        end_value: The id or token that is to be placed at the end of each
+            input segment (called "[SEP]" for BERT). The dtype much mach the
+            dtype of the input tensors to the layer.
+        pad_value: The id or token that is to be placed into the unused
+            unused positions after the last segment in the sequence
+            (called "[PAD]" for BERT).
+        truncator: The algorithm to truncate a list of batched segments to fit a
+            per-example length limit. The value can be either `round_robin` or
+            `waterfall`:
+                - `"round_robin"`: Available space is assigned one token at a
+                    time in a round-robin fashion to the inputs that still need
+                    some, until the limit is reached.
+                - `"waterfall"`: The allocation of the budget is done using a
+                    "waterfall" algorithm that allocates quota in a
+                    left-to-right manner and fills up the buckets until we run
+                    out of budget. It support arbitrary number of segments.
+
+    Examples:
+
+    *Pack a single input for classification.*
+    >>> seq1 = [1, 2, 3, 4]  # Replace with a real input.
+    >>> packer = keras_nlp.layers.BertPacker(8, start_value=101, end_value=102)
+    >>> packer(seq1)
+    {
+        "tokens": <tf.Tensor: shape=(8,), dtype=int32,
+            numpy=array([101,   1,   2,   3,   4, 102,   0,   0], dtype=int32)>,
+        "padding_mask": <tf.Tensor: shape=(8,), dtype=float32,
+            numpy=array([1., 1., 1., 1., 1., 1., 0., 0.], dtype=float32)>,
+        "segment_ids": <tf.Tensor: shape=(8,), dtype=float32,
+            numpy=array([0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32)>,
+    }
+
+    *Pack multiple inputs for classification.*
```

Comment:
How about I just show this with two sequences, that's the main use case.

This could actually be used for pretraining, the original BERT code just does in fact do "round robin", it just also truncates randomly from the front or back of a sequence to "add more randomness and avoid biases":
https://github.com/google-research/bert/blob/eedf5716ce1268e56f0a50264a88cafad334ac61/create_pretraining_data.py#L418

Reply:
Yes, an example with 2 sequences would resolve the confusion, thanks!

Reply:
sweet! i added that already